File: examples/nodejs/bankid.js
const axiosLib = require('axios');
const fs = require('fs');
const https = require('https');
const to = require('await-to-js').to;

const configs = {
  prod: {
    mobileBankIdPolicy: '1.2.752.78.1.5',
    bankdIdUrl: 'https://appapi2.bankid.com/rp/v5',
    pfx: 'YOUR PRODUCTION CERT INFO GOES HERE',
    passphrase: 'YOUR PRODUCTION CERT INFO GOES HERE',
    ca: fs.readFileSync(`./cert/prod.ca`),
  },
  test: {
    mobileBankIdPolicy: '1.2.3.4.25',
    bankdIdUrl: 'https://appapi2.test.bankid.com/rp/v5',
    pfx: fs.readFileSync('./cert/FPTestcert2_20150818_102329.pfx'),
    passphrase: 'qwerty123',
    ca: fs.readFileSync(`./cert/test.ca`),
  }
};

const config = configs.test;

// Create an axios instance that's configured to use the appropriate
// certificates and always send the required headers
const axios = axiosLib.create({
  httpsAgent: new https.Agent({
    pfx: config.pfx,
    passphrase: config.passphrase,
    ca: config.ca,
  }),
  headers: {
    'Content-Type': 'application/json',
  },
});

// This is a wrapper that will be used for all calls to BankID
async function call (method, params) {
  const [error, result] = await to(
    axios.post(`${config.bankdIdUrl}/${method}`, params));

  if (error) {
    // You will want to implement your own error handling here
    console.error('Error in call');
    if (error.response && error.response.data && error.response.data.errorCode === 'alreadyInProgress') {
      console.error(error.response.data);
      console.error('You would have had to call cancel on this orderRef before retrying');
      console.error('The order should now have been automatically cancelled by this premature retry');
    }
    return {error};
  }

  return result.data;
}

// BankID method call auth
const auth = async (personalNumber, endUserIp, otherDevice) =>
  await call('auth', {
    endUserIp,
    personalNumber,
    requirement: {
      allowFingerprint: true,
      ...otherDevice ? {certificatePolicies: [
        mobileBankIdPolicy,
      ]} : {},
    },
  });

// BankID method call sign
const sign = async (personalNumber, text, endUserIp, otherDevice) =>
  await call('sign', {
    endUserIp,
    personalNumber,
    userVisibleData: Buffer.from(text).toString('base64'),
    requirement: {
      allowFingerprint: true,
      ...otherDevice ? {certificatePolicies: [
        '1.2.752.78.1.5'
      ]} : {},
    },
  });

// BankID method call collect
const collect = async orderRef => await call('collect', {orderRef});

// BankID method call cancel
const cancel = async orderRef => await call('cancel', {orderRef});

// Returns the appropriate urls to launch the native client on mobile devices
// Note that all non-iOS devices should use `url`, while iOS has its own url
const launchUrls = autoStartToken =>
  (launchParams => ({
      url: `bankid://${launchParams}`,
      iosUrl: `https://app.bankid.com${launchParams}`,
    })
  )(`/?autostarttoken=[${autoStartToken}]&redirect=null`);

const persistResult = bankIdResult => 
  // FIXME - YOU MUST STORE:
  //   bankIdResult.signature
  //   bankIdResult.user
  //   bankIdResult.ocspResponse
  true;

const authFlow = async (pnr, endUserIp, launchFn) =>
  new Promise(async (resolve, reject) => {
    const {autoStartToken, orderRef} = await auth(pnr, endUserIp);
    if (!autoStartToken || !orderRef) {
      reject(new Error('Auth request failed'));
      return;
    }

    launchFn && launchFn({...launchUrls(autoStartToken), orderRef});
  
    const interval = setInterval(async () => {
      const {status, hintCode, completionData} = await collect(orderRef);
      if (status === 'failed') {
        clearInterval(interval);
        resolve({ok: false, status: hintCode});
      } else if (status === 'complete') {
        clearInterval(interval);
        resolve({ok: true, status: completionData});
        persistResult(completionData);
      } else {
        console.log(hintCode);
      }
    }, 2000);
  });

const signFlow = async (pnr, endUserIp, text, launchFn) =>
  new Promise(async (resolve, reject) => {
    const {autoStartToken, orderRef} = await sign(pnr, endUserIp, text);
    if (!autoStartToken || !orderRef) {
      reject(new Error('Sign request failed'));
      return;
    }

    launchFn && launchFn({...launchUrls(autoStartToken), orderRef});
  
    const interval = setInterval(async () => {
      const {status, hintCode, completionData} = await collect(orderRef);
      if (status === 'failed') {
        clearInterval(interval);
        resolve({ok: false, status: hintCode});
      } else if (status === 'complete') {
        clearInterval(interval);
        resolve({ok: true, status: completionData});
        persistResult(completionData);
      } else {
        console.log(hintCode);
      }
    }, 2000);
  });

const runningAsScript = !module.parent;

if (runningAsScript) {
  function usage () {
    console.log('Usage: bankid [operation] [param]');
    console.log('       operation can be one of (auth, sign, cancel)');
    console.log('       for auth and sign, param must be a Swedish personnummer');
    console.log('       for cancel, param must be an orderRef previously started by auth or sign');
  }

  if (process.argv.length < 4) {
    usage();
    process.exit(1);
  }

  const operation = process.argv[2];
  const param = process.argv[3];

  function launchNativeApp ({url, iosUrl, orderRef}) {
    console.log(`Flow started with orderRef ${orderRef}`);
    console.log(`FIXME: launch native app on ${url} or ${iosUrl}`);
  }

  const dummyIp = '127.0.0.1';
  
  async function testAuth (pnr) {
    const result = await authFlow(pnr, dummyIp, launchNativeApp);
    console.log(result.ok, result.status.user);
  }
  
  async function testSign (pnr) {
    const result = await signFlow(pnr, dummyIp, 'Test text for signing', launchNativeApp);
    console.log(result.ok, result.status.user);
  }

  ({
    auth: testAuth,
    sign: testSign,
    cancel: cancel,
  }[operation] || (_ => console.log(`Unknown operation ${operation}`)))(param);
}

module.exports = {
  authFlow,
  signFlow,
  cancel,
};

File: examples/python/bankid.py
import sys
import time, threading
import base64
import tempfile
import requests
import OpenSSL.crypto
import contextlib
import urllib3
import json

urllib3.disable_warnings()

CONFIGS = {
    'prod': {
        'mobileBankIdPolicy': '1.2.752.78.1.5',
        'bankIdUrl': 'https://appapi2.bankid.com/rp/v5',
        'pfx': 'YOUR PRODUCTION CERT INFO GOES HERE',
        'passphrase': 'YOUR PRODUCTION CERT INFO GOES HERE',
        'ca': './cert/test.ca'
    },
    'test': {
        'mobileBankIdPolicy': '1.2.3.4.25',
        'bankIdUrl': 'https://appapi2.test.bankid.com/rp/v5',
        'pfx': './cert/FPTestcert2_20150818_102329.pfx',
        'passphrase': 'qwerty123',
        'ca': './cert/test.ca'
    }
}

config = CONFIGS['test']

@contextlib.contextmanager
def pfx_to_pem(pfx_path, pfx_password):
    ''' Decrypts the .pfx file to be used with requests. '''
    with tempfile.NamedTemporaryFile(suffix='.pem') as t_pem:
        f_pem = open(t_pem.name, 'wb')
        pfx = open(pfx_path, 'rb').read()
        p12 = OpenSSL.crypto.load_pkcs12(pfx, pfx_password)
        f_pem.write(OpenSSL.crypto.dump_privatekey(OpenSSL.crypto.FILETYPE_PEM, p12.get_privatekey()))
        f_pem.write(OpenSSL.crypto.dump_certificate(OpenSSL.crypto.FILETYPE_PEM, p12.get_certificate()))
        ca = p12.get_ca_certificates()
        if ca is not None:
            for cert in ca:
                f_pem.write(OpenSSL.crypto.dump_certificate(OpenSSL.crypto.FILETYPE_PEM, cert))
        f_pem.close()
        yield t_pem.name

class SetInterval:
    def __init__(self, interval, action):
        self.interval = interval
        self.action = action
        self.stopEvent = threading.Event()
        thread = threading.Thread(target = self.__set_interval)
        thread.start()

    def __set_interval(self):
        nextTime = time.time() + self.interval
        while not self.stopEvent.wait(nextTime - time.time()) :
            nextTime += self.interval
            self.action()

    def cancel(self):
        self.stopEvent.set()

def call(method, params):
    with pfx_to_pem(config['pfx'], config['passphrase']) as cert:
        r = requests.post(config['bankIdUrl'] + '/' + method, cert=cert, json=params, verify=config['ca'])

    return r.json()

def auth(pnr, end_user_ip, other_device = False) :
    requirement = {
        'allowFingerprint': True,
    }
    if other_device:
        requirement['certificatePolicies'] = [config['mobileBankIdPolicy']]

    return call('auth', {
        'endUserIp': end_user_ip,
        'personalNumber': pnr,
        'requirement': requirement
    })

def sign(pnr, end_user_ip, text, other_device = False):
    requirement = {
        'allowFingerprint': True,
    }
    if other_device:
        requirement['certificatePolicies'] = [config['mobileBankIdPolicy']]

    return call('sign', {
        'endUserIp': end_user_ip,
        'personalNumber': pnr,
        'userVisibleData': base64.b64encode(text.encode()).decode(),
        'requirement': requirement
    })

def collect(order_ref):
    return call('collect', {
        'orderRef': order_ref
    })

def cancel(order_ref):
    return call('cancel', {
        'orderRef': order_ref
    })

def launch_urls(auto_start_token):
    launch_params = '/?autostarttoken=[' + auto_start_token + ']&redirect=null'
    return {
      'url': 'bankid://' + launch_params,
      'iosUrl': 'https://app.bankid.com' + launch_params,
    }

def persist_result(bank_id_result):
    # FIXME - YOU MUST STORE:
    #   bank_id_result.signature
    #   bank_id_result.user
    #   bank_id_result.ocspResponse
    pass

def auth_flow(pnr, end_user_ip, launch_fn):
    r = auth(pnr, end_user_ip)
    if not 'autoStartToken' in r or not 'orderRef' in r:
        print(r)
        raise ValueError('Auth request failed')

    auto_start_token = r['autoStartToken']
    order_ref = r['orderRef']

    if launch_fn:
        urls = launch_urls(auto_start_token)
        launch_fn({
            'url': urls['url'],
            'iosUrl': urls['iosUrl'],
            'orderRef': order_ref
        })

    def poll():
        r = collect(order_ref)
        if r['status'] == 'failed':
            interval.cancel()
            print('Authorization failed')
            print(r['hintCode'])
        elif r['status'] == 'complete':
            interval.cancel()
            persist_result(r)
            print('Authorization successful')
            print(r['completionData']['user'])
            print(r['completionData']['device'])
        else:
            print(r['hintCode'])

    interval = SetInterval(2, poll)

def sign_flow(pnr, end_user_ip, text, launch_fn):
    r = sign(pnr, end_user_ip, text)
    if not 'autoStartToken' in r or not 'orderRef' in r:
        print(r)
        raise ValueError('Sign request failed')

    auto_start_token = r['autoStartToken']
    order_ref = r['orderRef']

    if launch_fn:
        urls = launch_urls(auto_start_token)
        launch_fn({
            'url': urls['url'],
            'iosUrl': urls['iosUrl'],
            'orderRef': order_ref
        })

    def poll():
        r = collect(order_ref)
        if r['status'] == 'failed':
            interval.cancel()
            print('Signing failed')
            print(r['hintCode'])
        elif r['status'] == 'complete':
            interval.cancel()
            persist_result(r)
            print('Signing successful')
            print(r['completionData']['user'])
            print(r['completionData']['device'])
        else:
            print(r['hintCode'])

    interval = SetInterval(2, poll)

def usage():
    print('Usage: %s [operation] [param]' % sys.argv[0])
    print('       operation can be one of (auth, sign, cancel)')
    print('       for auth and sign, param must be a Swedish personnummer')
    print('       for cancel, param must be an orderRef previously started by auth or sign')
    sys.exit()

if len(sys.argv) < 3:
    usage()

operation = sys.argv[1]
if operation not in ['auth', 'sign', 'cancel']:
    usage()

pnr_or_order_ref = sys.argv[2]

launch_native_app = lambda launch_info: print('TODO: Launch native app ' + json.dumps(launch_info))

try:
    if operation == 'auth':
        auth_flow(pnr_or_order_ref, '123.123.123.123', launch_native_app)
    elif operation == 'sign':
        sign_flow(pnr_or_order_ref, '123.123.123.123', 'Test text for signing', launch_native_app)
    elif operation == 'cancel':
        cancel(pnr_or_order_ref)
except ValueError as e:
    print(e)
